rasterio/__init__.py

```diff
@@ -280,7 +280,7 @@ def open(
 
         # Note: FilePath does not support writing and an exception will
         # result from this.
-        elif mode.startswith("w"):
+        else:
```

Comment:
Because of the constraint above, this can be `else`, with the benefit that we always return or raise an exception after entering into the branch at line 277.